Automattic/wp-calypso
# Jetpack User Licensing: Add redux state management

#### Changes proposed in this Pull Request

This PR implements redux state management for 'user' type licenses and licenses counts.

- Added the reducer, actions, and selectors for 'user' type licenses.
- This adds a `userLicensing` property to the global state object.
- Added a `<QueryJetpackUserLicenses />` component for fetching users licenses.
- Added various different selector functions for retrieving specific user licenses data.
- Added unit tests for the reducer, actions, and selectors.

Related to 1201096622142517-as-1201096624355390

#### Testing instructions

- Checkout this branch and `yarn start`.
- Verify all unit tests are passing:
    - Run command: `yarn test-client client/state/user-licensing`
- Inspect all the code, verify it looks clean and reasonable.

Query the user licenses data:

Prerequisites:
- [Redux DevTools](https://chrome.google.com/webstore/detail/redux-devtools/lmhkpmbekcpmknklioeibfkpmmfibljd) installed into your browser.
- Ideally you should have 1 or more Jetpack user licenses. (If you don't have one, you can still test, however you won't have any data in the `licenses` and `counts` properties. To create a user license, 1. you can log into your wpcom sandbox and create on in wp-shell(`wpsh`), instructions here: D68384-code. Or 2. you can enable a flag(`add_action`) in your sandbox and complete checkout with a jetpack product, instructions here: D67941-code).

Instructions:
- Open file: `client/my-sites/plans/jetpack-plans/selector.tsx`;
- Right after line 6, add `import QueryJetpackUserLicenses from 'calypso/components/data/query-jetpack-user-licenses';`
- In the top of the `render()` function, add component: `<QueryJetpackUserLicenses />`  (just above the `<QueryJetpackSaleCoupon />` component).
- Go to: http://calypso.localhost:3000/plans/:site`  (replace `:site` with the site slug of any Jetpack site you own).
- Open your browser developer tools and open Redux DevTools. Inspect Calypso state and find `state.userLicensing`. Inspect the object and all the data. Verify it looks all good. (See screenshots below)
- 
<img width="821" alt="Screenshot on 2021-10-19 at 13-03-31" src="https://user-images.githubusercontent.com/11078128/137958115-28c7b2b4-f44f-42b9-b998-fc9c3942201c.png">

![Screenshot on 2021-10-19 at 13-06-03](https://user-images.githubusercontent.com/11078128/137958297-7d1fe8ec-c0e1-469e-9dd3-6a00da7ec544.png)
